drivers: uart: nrf: fixing interrupt driven API

jakub-uC opened the pull request:
Implemented API functions: uart_nrfx_irq_tx_complete
and uart_nrfx_irq_tx_ready were returning not correct
values.

1. uart_nrfx_irq_tx_complete was returning wrong
value once first byte was send.
2. uart_nrfx_irq_tx_ready was returning wrong
value before first byte was send.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

carlescufi commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -15,6 +15,7 @@
 
 #ifdef CONFIG_UART_INTERRUPT_DRIVEN
 static uart_irq_callback_t m_irq_callback; /**< Callback function pointer */
+static bool m_init_uart_event_txrdy;
```
we don't use hungarian notation prefixes in Zephyr. Can you please remove the `m_` from this new variable and also from the `m_irq_callback` (in a separate commit for that one please).

carlescufi replied:
Another question, does this `bool` need to be atomic? we have `atomic_t` for this

carlescufi commented:
@jarz-nordic thanks for the PR. Looks good to me apart from a couple of things I noted down in the review.

ioannisg reviewed (changes requested):
requires, rework

ioannisg commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -150,6 +151,11 @@ static unsigned char uart_nrfx_poll_out(struct device *dev,
 	/* send a character */
 	nrf_uart_txd_set(NRF_UART0, (u8_t)c);
 
+	/* First byte was sent, from now on relay on NRF_UART_EVENT_TXDRDY event
```
The second part of the sentence-comment is slightly confusing: what does 

> relay on NRF_UART_EVENT_TXDRDY event status

 mean?

ioannisg commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -180,6 +186,11 @@ static int uart_nrfx_fifo_fill(struct device *dev,
 	u8_t num_tx = 0;
 
+	/* First byte was sent, from now on relay on NRF_UART_EVENT_TXDRDY event
+	 * status.
```
Same comment as above

ioannisg commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -395,6 +410,15 @@ static int uart_nrfx_init(struct device *dev)
 
 #ifdef CONFIG_UART_INTERRUPT_DRIVEN
 
+	/* Global variable: m_init_uart_event_txrdy is necessary to make
```
This comment looks really confusing, IMHO. I don't get what the "correct value" is or shall be.

ioannisg commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -395,6 +410,15 @@ static int uart_nrfx_init(struct device *dev)
+	/* Global variable: m_init_uart_event_txrdy is necessary to make
+	 * API functions: uart_nrfx_irq_tx_complete and uart_nrfx_irq_tx_ready
+	 * return always correct value.
+	 * These API functions will return:
```
Could we just, rephrase, the following lines, to be more "to the point" ? E.g.
"The aforementioned API functions will always return true before the driver has transmitted any characters, regardless of the status of the TX_READY event flag. After that the functions will always return NRF_UART_EVENT_RXDRDY after the driver has transmitted the first character".

jarz-nordic replied:
Sure I can use it.

ioannisg commented:
@jarz-nordic , thanks for the PR; I left some comments to think about. However, I am just wondering: could we, simply, clear the EVENTS TX_READY upon initialization and _not_ introduce the state variable? It looks to me the state variable is only perform a state transition _once_.

jarz-nordic commented:
@ioannisg : NRF_UART_EVENT_TXDRDY event is set once first byte is shifted out from TXD register and we are allowed to send next byte. As a result NRF_UART_EVENT_TXDRDY event is used by functions: uart_nrfx_irq_tx_complete and uart_nrfx_irq_tx_ready to indicate that nothing remains to be transmited and we are ready to send next byte respectively.

However just after peripheral intialization event NRF_UART_EVENT_TXDRDY is not set and cannot be used by these functions. To fix that I am using global variable which is set to true at the initialization time and reset after first byte shifted out.

ioannisg commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -14,7 +14,10 @@
-static uart_irq_callback_t m_irq_callback; /**< Callback function pointer */
+
+static uart_irq_callback_t irq_callback; /**< Callback function pointer */
+static volatile u8_t init_uart_event_txrdy;
```
Some description of the state variable is required, here.

jarz-nordic commented:
This PR require some more work. Apperantly interrupt driven UART has not been tested yet.

Function: uart_nrfx_irq_tx_enable apart from enabling an interrupt shall also **trigger it!** what is not happening now.

I will add one more commit to this PR, but it will be not beautiful as our peripheral does not provide easy solution for this.

carlescufi commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -215,6 +238,18 @@ static int uart_nrfx_fifo_read(struct device *dev,
+	/* Critical section is used to avoid any UART related interrupt which
+	 * can occure after the if statement and before call of the function
+	 * forcing an interrupt. */
+	irq_lock();
```
You need to store the key you get with `irq_lock()` and provide it to `irq_unlock()` later

jarz-nordic replied:
fixed

carlescufi commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -215,6 +238,18 @@ static int uart_nrfx_fifo_read(struct device *dev,
+	/* Critical section is used to avoid any UART related interrupt which
+	 * can occure after the if statement and before call of the function
+	 * forcing an interrupt. */
+	u32_t key = irq_lock();
```
Can you please place the variable declaration in the beginning of the function? This is not explicitly stated, but it's what we've been doing. Thanks!

ioannisg commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -14,7 +14,12 @@
+
+static uart_irq_callback_t irq_callback; /**< Callback function pointer */
+/* Help variable used by some functions to return correct driver status before
+ * first byte is shifted out from UART0. */
```
Strictly speaking, this is not specific to UART"0"

jarz-nordic replied:
it is specific to UART0 because in all our chips we have either UART0 only or no UART at all, only UARTE.
For UARTE there is simpler warkaround.

ioannisg replied:
We have UART1 in Graviton, at least.
I know that the driver, now, uses UART0, but in theory, we could use UART1 as well. The comment could cover all. So "UART" is better, I think.

ioannisg commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -150,6 +155,16 @@ static unsigned char uart_nrfx_poll_out(struct device *dev,
+	 * From now on event flag: NRF_UART_EVENT_TXDRDY status can be used
+	 * to indicate that nothing remains to be transmited and peripheral
+	 * is ready to send next byte.
+	 */
```
nice

jarz-nordic commented:
@anangl : May you please review it as well?

evenl commented:
Tested interrupt driven UART reception and transmission with this patch on a nrf52840_pca10056 board, this code fixes the described issues.

ioannisg commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -214,7 +253,24 @@ static int uart_nrfx_fifo_read(struct device *dev,
 	nrf_uart_int_enable(NRF_UART0, NRF_UART_INT_MASK_TXDRDY);
+
+	/* Critical section is used to avoid any UART related interrupt which
+	 * can occure after the if statement and before call of the function
```
occur*

jarz-nordic replied:
fixed

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -180,13 +207,25 @@ static int uart_nrfx_fifo_fill(struct device *dev,
 	while ((len - num_tx > 0) &&
-	       nrf_uart_event_check(NRF_UART0, NRF_UART_EVENT_TXDRDY)) {
+	       (nrf_uart_event_check(NRF_UART0, NRF_UART_EVENT_TXDRDY) ||
+	        uart_sw_event_txrdy)) {
```
This (together with these two changes above and one below) becomes hard to maintain and consequently, error prone. I suggest to introduce two functions as follows:
```
static bool event_txdrdy_check(void)
{
	return (nrf_uart_event_check(NRF_UART0, NRF_UART_EVENT_TXDRDY)
#ifdef CONFIG_UART_INTERRUPT_DRIVEN
		|| uart_sw_event_txdrdy
#endif
	       );
}

static void event_txdrdy_clear(void)
{
	nrf_uart_event_clear(NRF_UART0, NRF_UART_EVENT_TXDRDY);
#ifdef CONFIG_UART_INTERRUPT_DRIVEN
	uart_sw_event_txdrdy = 0;
#endif
}
```
and just use them everywhere in this file instead of `nrf_uart_event_check(NRF_UART0, NRF_UART_EVENT_TXDRDY)` and `nrf_uart_event_clear(NRF_UART0, NRF_UART_EVENT_TXDRDY)`, respectively.

And in the comment before the `uart_sw_event_txdrdy` variable definition you can just write that this variable is used to simulate the TXDRDY event before the first byte is sent, so that this event can be used as an indication of the transmitter status.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -14,7 +14,19 @@
+/* Help variable used by some functions to return correct driver status before
+ * first byte is shifted out from UART0. 
+ */
+static volatile u8_t uart_sw_event_txrdy;
```
It should be named `uart_sw_event_txdrdy` to follow the real event's name.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -395,6 +454,14 @@ static int uart_nrfx_init(struct device *dev)
+	/* The aforementioned API functions will always return true before
+	 * the driver has transmitted any characters, regardless of the status
+	 * of the TX_READY event flag. After that the functions will always
+	 * return NRF_UART_EVENT_RXDRDY after the driver has transmitted
```
This comment should be rephrased. I don't understand which "aforementioned" functions you refer to, what is the "TX_READY event flag", and what `NRF_UART_EVENT_RXDRDY` has to do with all of this.

jarz-nordic commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -14,7 +14,19 @@
+
+/* Help variable used to indicate that software forced interrupt is pending
+ */
+static volatile u8_t uart_sw_event_txrdy_pending;
```
Because pending status can be only used when interrupts are enabled, while uart_sw_event_txrdy is true already after initialization

jarz-nordic commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -244,7 +302,9 @@ static void uart_nrfx_irq_rx_disable(struct device *dev)
-	return !nrf_uart_event_check(NRF_UART0, NRF_UART_EVENT_TXDRDY);
+	int ret = uart_sw_event_txrdy ||
+		nrf_uart_event_check(NRF_UART0, NRF_UART_EVENT_TXDRDY);
+	return ret;
```
ok

jarz-nordic commented:
done

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -387,14 +427,21 @@ static int uart_nrfx_init(struct device *dev)
+	 * to send data.
+	 * Once first byte is send out readiness status will be provided by
+	 * hardware event: NRF_UART_EVENT_TXDRDY.
+	 */
```
What about writing just "Simulate that the TXDRDY event is set, so that the transmitter status is indicated correctly."?

jarz-nordic replied:
done

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -407,24 +454,41 @@ static int uart_nrfx_init(struct device *dev)
-	.poll_in          = uart_nrfx_poll_in,          /** Console I/O function */
-	.poll_out         = uart_nrfx_poll_out,         /** Console I/O function */
-	.err_check        = uart_nrfx_err_check,        /** Console I/O function */
+	/** Console I/O function */
```
What was wrong with the previous layout?
Including such changes in the same commit obscures the actual modifications that the commit was intended to introduce.

jarz-nordic replied:
I was afraid that I have edited a line which exceed 80 characters span. I will remove these comments.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -407,24 +454,41 @@ static int uart_nrfx_init(struct device *dev)
+	/** IRQ receiver disabling function */
+	.irq_rx_disable   = uart_nrfx_irq_rx_disable,
+	/** IRQ transfer complete function */
+	.irq_tx_complete  = uart_nrfx_irq_tx_ready,
```
This looks confusing (seems to be a mistake).
It should be at least commented (that a common function is used for two API entries, since our hardware don't provide these two statuses separately). And I would suggest to include `_complete` in the function name as well.

anangl reviewed (changes requested):
You need to distribute the changes properly among the commits. Currently, there is some mess in this.

jarz-nordic commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -387,14 +427,21 @@ static int uart_nrfx_init(struct device *dev)
 	nrf_uart_enable(NRF_UART0);
 
-	nrf_uart_event_clear(NRF_UART0, NRF_UART_EVENT_TXDRDY);
+	event_txdrdy_clear();
```
done

jarz-nordic replied:
good point

jarz-nordic commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -14,9 +14,41 @@
+	nrf_uart_event_clear(NRF_UART0, NRF_UART_EVENT_TXDRDY);
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	/* First byte was sent and uart_sw_event_txdrdy variable is not needed
+	 * any longer.
```
done

Commit:
drivers: uart: nrf: fixing interrupt driven API

1. API function: uart_nrfx_irq_tx_enable is expected to enable and
trigger TXDRDY interrupt. Due to HW limiation before first byte has
been sent TXDRDY interrupt will not be triggered so it must be forced
by the software.

2. Common function: uart_nrfx_irq_tx_ready_complete is used for two
API entries: irq_tx_ready, irq_tx_complete because Nordic hardware does
not distinguish between them. This function returns 1 when driver is
ready to accept data and 0 otherwise.

3. Removed Hungarian prefixes.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

nvlsianpu commented:
@jarz-nordic @carlescufi I Tested this along with mcuboot serial recovery - it worked well with this patch.

carlescufi commented:
thanks for the testing @nvlsianpu

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -14,9 +14,36 @@
+
+/* Variable used to override the state of the TXDRDY event in the initial state
+ * of the driver. Before any byte is sent.the event is not set and we want
+ * to use it as an indication if the transmitter is ready to accept a new byte.
```
"Before ..." -> "This event is not set by the hardware until a first byte is sent, and we want ..."

anangl replied:
Only this one correction needed. I can live with the other changes. ;-)

jarz-nordic replied:
done
